server/src/test/java/org/elasticsearch/cluster/FeatureAwareTests.java

```diff
@@ -0,0 +1,172 @@
+import org.elasticsearch.test.VersionUtils;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.Optional;
+
+import static org.elasticsearch.test.VersionUtils.randomVersionBetween;
+
+public class FeatureAwareTests extends ESTestCase {
+
+    abstract static class Custom implements MetaData.Custom {
+
+        private final Version version;
+
+        Custom(final Version version) {
+            this.version = version;
+        }
+
+        @Override
+        public EnumSet<MetaData.XContentContext> context() {
+            return MetaData.ALL_CONTEXTS;
+        }
+
+        @Override
+        public Diff<MetaData.Custom> diff(final MetaData.Custom previousState) {
+            return null;
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+
+        }
+
+        @Override
+        public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+            return builder;
+        }
+
+        @Override
+        public Version getMinimalSupportedVersion() {
+            return version;
+        }
+
+    }
+
+    static class NoRequiredFeatureCustom extends Custom {
+
+        NoRequiredFeatureCustom(final Version version) {
+            super(version);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return "no-required-feature";
+        }
+
+    }
+
+    static class RequiredFeatureCustom extends Custom {
+
+        RequiredFeatureCustom(final Version version) {
+            super(version);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return null;
+        }
+
+        @Override
+        public Optional<String> getRequiredFeature() {
+            return Optional.of("required-feature");
+        }
+
+    }
+
+    public void testVersion() {
+        final Version version = VersionUtils.randomVersion(random());
+        for (final Custom custom : Arrays.asList(new NoRequiredFeatureCustom(version), new RequiredFeatureCustom(version))) {
+            {
+                final BytesStreamOutput out = new BytesStreamOutput();
+                final Version afterVersion = randomVersionBetween(random(), version, Version.CURRENT);
+                out.setVersion(afterVersion);
+                if (custom.getRequiredFeature().isPresent()) {
+                    out.setFeatures(Collections.singleton(custom.getRequiredFeature().get()));
+                }
+                assertTrue(FeatureAware.shouldSerializeCustom(out, custom));
+            }
+            {
+                final BytesStreamOutput out = new BytesStreamOutput();
+                final Version beforeVersion =
+                        randomVersionBetween(random(), VersionUtils.getFirstVersion(), VersionUtils.getPreviousVersion(version));
+                out.setVersion(beforeVersion);
```

Comment:
randomly add the required feature and tit shouldn't matter?

Reply:
I pushed 8be176c24b6aa9e838e51b7c4a1ee0daeffaaecb.